enable bf16 embeddingbag with fbgemm fast path in aten

zhuhaozhe opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #91949
* #89199

Summary:

This PR adds BF16 support For embedding_bag(CPU) op with FBGEMM kernel
(depends on https://github.com/pytorch/FBGEMM/pull/1540)
 
Benchmark result on CPX 28 core/socket BS = 28 * 128:
```python
import torch

a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)

def cache_flush():
    # We assume the cache size is <= 512MB here.
    # a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # a, b are initialized out of this function to avoid allocate memory every time
    global a, b
    a += b

BATCHSIZE = 128 * 28

E = torch.nn.EmbeddingBag(100000, 512)

import time

for BAGSIZE in [20, 10, 1]:
    input = torch.randint(0, 100000, (BATCHSIZE * BAGSIZE, ))
    offset = torch.arange(0, BATCHSIZE * BAGSIZE, BAGSIZE)
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = E(input, offset)
        t +=time.time() - start
    print("fp32:", t)

    bf16 = E.bfloat16()
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = bf16(input, offset)
        t +=time.time() - start
    print("bf16:",t)
```


| BAG SIZE | FP32               | BF16                | ratio | flush cache |
|----------|--------------------|---------------------|-------|-------------|
| 1        | 0.42674922943115234| 0.4264225959777832| 1.0007659853311013| TRUE        |
| 10       | 1.0786781311035156| 1.0770792961120605| 1.0014844171615094| TRUE        |
| 20       | 2.4907093048095703  | 1.612083911895752| 1.545024602274324| TRUE        |
| 1        | 0.12787652015686035 | 0.12776947021484375| 1.0008378366274557| FALSE       |
| 10       | 0.40375709533691406 | 0.4228525161743164  | 0.9548414160800914  | FALSE       |
| 20       | 2.0639474391937256 | 0.9705793857574463  | 2.129852698601417  | FALSE       |

cc @jgong5 @mingfeima @XiaobingSuper @sanchitintel @ashokei @jingxu10

zhuhaozhe committed:
enable bf16 embeddingbag with fbgemm fast path in aten

[ghstack-poisoned]

zhuhaozhe committed:
Update on "enable bf16 embeddingbag with fbgemm fast path in aten"


Summary:

This PR adds BF16 support For embedding_bag(CPU) op with FBGEMM kernel
(depends on https://github.com/pytorch/FBGEMM/pull/1540)
 
Benchmark result on CPX 28 core/socket BS = 28 * 128:
```python
import torch

a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)

def cache_flush():
    # We assume the cache size is <= 512MB here.
    # a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # a, b are initialized out of this function to avoid allocate memory every time
    global a, b
    a += b

BATCHSIZE = 128 * 28

E = torch.nn.EmbeddingBag(100000, 512)

import time

for BAGSIZE in [20, 10, 1]:
    input = torch.randint(0, 100000, (BATCHSIZE * BAGSIZE, ))
    offset = torch.arange(0, BATCHSIZE * BAGSIZE, BAGSIZE)
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = E(input, offset)
        t +=time.time() - start
    print("fp32:", t)

    bf16 = E.bfloat16()
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = bf16(input, offset)
        t +=time.time() - start
    print("bf16:",t)
```


| BAG SIZE | FP32               | BF16                | ratio | flush cache |
|----------|--------------------|---------------------|-------|-------------|
| 1        | 0.42674922943115234| 0.4264225959777832| 1.0007659853311013| TRUE        |
| 10       | 1.0786781311035156| 1.0770792961120605| 1.0014844171615094| TRUE        |
| 20       | 2.4907093048095703  | 1.612083911895752| 1.545024602274324| TRUE        |
| 1        | 0.12787652015686035 | 0.12776947021484375| 1.0008378366274557| FALSE       |
| 10       | 0.40375709533691406 | 0.4228525161743164  | 0.9548414160800914  | FALSE       |
| 20       | 2.0639474391937256 | 0.9705793857574463  | 2.129852698601417  | FALSE       |

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

jgong5 commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -281,10 +267,43 @@ index_select_add(
           });
-      return;
+    } else {
+      using bfloat16 = int16_t;
```
seems the code is very like the float16 path, can we reuse?

zhuhaozhe replied:
We have to use uint16_t as fbgemm::bfloat16, and add an arg `isbf16` to distinguish fp16/bf16. Now the code here is reused better.

jgong5 commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -663,10 +681,47 @@ index_select_scale_add(
-      return;
+    } else {
+      using bfloat16 = int16_t;
+      fbgemm::Bfloat16ToFloat_simd(
```
ditto: code reuse float16 path?

zhuhaozhe committed:
Update on "enable bf16 embeddingbag with fbgemm fast path in aten"


Summary:

This PR adds BF16 support For embedding_bag(CPU) op with FBGEMM kernel
(depends on https://github.com/pytorch/FBGEMM/pull/1540)
 
Benchmark result on CPX 28 core/socket BS = 28 * 128:
```python
import torch

a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)

def cache_flush():
    # We assume the cache size is <= 512MB here.
    # a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # a, b are initialized out of this function to avoid allocate memory every time
    global a, b
    a += b

BATCHSIZE = 128 * 28

E = torch.nn.EmbeddingBag(100000, 512)

import time

for BAGSIZE in [20, 10, 1]:
    input = torch.randint(0, 100000, (BATCHSIZE * BAGSIZE, ))
    offset = torch.arange(0, BATCHSIZE * BAGSIZE, BAGSIZE)
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = E(input, offset)
        t +=time.time() - start
    print("fp32:", t)

    bf16 = E.bfloat16()
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = bf16(input, offset)
        t +=time.time() - start
    print("bf16:",t)
```


| BAG SIZE | FP32               | BF16                | ratio | flush cache |
|----------|--------------------|---------------------|-------|-------------|
| 1        | 0.42674922943115234| 0.4264225959777832| 1.0007659853311013| TRUE        |
| 10       | 1.0786781311035156| 1.0770792961120605| 1.0014844171615094| TRUE        |
| 20       | 2.4907093048095703  | 1.612083911895752| 1.545024602274324| TRUE        |
| 1        | 0.12787652015686035 | 0.12776947021484375| 1.0008378366274557| FALSE       |
| 10       | 0.40375709533691406 | 0.4228525161743164  | 0.9548414160800914  | FALSE       |
| 20       | 2.0639474391937256 | 0.9705793857574463  | 2.129852698601417  | FALSE       |

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

zhuhaozhe committed:
Update on "enable bf16 embeddingbag with fbgemm fast path in aten"


Summary:

This PR adds BF16 support For embedding_bag(CPU) op with FBGEMM kernel
(depends on https://github.com/pytorch/FBGEMM/pull/1540)
 
Benchmark result on CPX 28 core/socket BS = 28 * 128:
```python
import torch

a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)

def cache_flush():
    # We assume the cache size is <= 512MB here.
    # a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # a, b are initialized out of this function to avoid allocate memory every time
    global a, b
    a += b

BATCHSIZE = 128 * 28

E = torch.nn.EmbeddingBag(100000, 512)

import time

for BAGSIZE in [20, 10, 1]:
    input = torch.randint(0, 100000, (BATCHSIZE * BAGSIZE, ))
    offset = torch.arange(0, BATCHSIZE * BAGSIZE, BAGSIZE)
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = E(input, offset)
        t +=time.time() - start
    print("fp32:", t)

    bf16 = E.bfloat16()
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = bf16(input, offset)
        t +=time.time() - start
    print("bf16:",t)
```


| BAG SIZE | FP32               | BF16                | ratio | flush cache |
|----------|--------------------|---------------------|-------|-------------|
| 1        | 0.42674922943115234| 0.4264225959777832| 1.0007659853311013| TRUE        |
| 10       | 1.0786781311035156| 1.0770792961120605| 1.0014844171615094| TRUE        |
| 20       | 2.4907093048095703  | 1.612083911895752| 1.545024602274324| TRUE        |
| 1        | 0.12787652015686035 | 0.12776947021484375| 1.0008378366274557| FALSE       |
| 10       | 0.40375709533691406 | 0.4228525161743164  | 0.9548414160800914  | FALSE       |
| 20       | 2.0639474391937256 | 0.9705793857574463  | 2.129852698601417  | FALSE       |

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

zhuhaozhe committed:
Update on "enable bf16 embeddingbag with fbgemm fast path in aten"


Summary:

This PR adds BF16 support For embedding_bag(CPU) op with FBGEMM kernel
(depends on https://github.com/pytorch/FBGEMM/pull/1540)
 
Benchmark result on CPX 28 core/socket BS = 28 * 128:
```python
import torch

a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)

def cache_flush():
    # We assume the cache size is <= 512MB here.
    # a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # a, b are initialized out of this function to avoid allocate memory every time
    global a, b
    a += b

BATCHSIZE = 128 * 28

E = torch.nn.EmbeddingBag(100000, 512)

import time

for BAGSIZE in [20, 10, 1]:
    input = torch.randint(0, 100000, (BATCHSIZE * BAGSIZE, ))
    offset = torch.arange(0, BATCHSIZE * BAGSIZE, BAGSIZE)
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = E(input, offset)
        t +=time.time() - start
    print("fp32:", t)

    bf16 = E.bfloat16()
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = bf16(input, offset)
        t +=time.time() - start
    print("bf16:",t)
```


| BAG SIZE | FP32               | BF16                | ratio | flush cache |
|----------|--------------------|---------------------|-------|-------------|
| 1        | 0.42674922943115234| 0.4264225959777832| 1.0007659853311013| TRUE        |
| 10       | 1.0786781311035156| 1.0770792961120605| 1.0014844171615094| TRUE        |
| 20       | 2.4907093048095703  | 1.612083911895752| 1.545024602274324| TRUE        |
| 1        | 0.12787652015686035 | 0.12776947021484375| 1.0008378366274557| FALSE       |
| 10       | 0.40375709533691406 | 0.4228525161743164  | 0.9548414160800914  | FALSE       |
| 20       | 2.0639474391937256 | 0.9705793857574463  | 2.129852698601417  | FALSE       |

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

zhuhaozhe committed:
Update on "enable bf16 embeddingbag with fbgemm fast path in aten"


Summary:

This PR adds BF16 support For embedding_bag(CPU) op with FBGEMM kernel
(depends on https://github.com/pytorch/FBGEMM/pull/1540)
 
Benchmark result on CPX 28 core/socket BS = 28 * 128:
```python
import torch

a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)

def cache_flush():
    # We assume the cache size is <= 512MB here.
    # a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # a, b are initialized out of this function to avoid allocate memory every time
    global a, b
    a += b

BATCHSIZE = 128 * 28

E = torch.nn.EmbeddingBag(100000, 512)

import time

for BAGSIZE in [20, 10, 1]:
    input = torch.randint(0, 100000, (BATCHSIZE * BAGSIZE, ))
    offset = torch.arange(0, BATCHSIZE * BAGSIZE, BAGSIZE)
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = E(input, offset)
        t +=time.time() - start
    print("fp32:", t)

    bf16 = E.bfloat16()
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = bf16(input, offset)
        t +=time.time() - start
    print("bf16:",t)
```


| BAG SIZE | FP32               | BF16                | ratio | flush cache |
|----------|--------------------|---------------------|-------|-------------|
| 1        | 0.42674922943115234| 0.4264225959777832| 1.0007659853311013| TRUE        |
| 10       | 1.0786781311035156| 1.0770792961120605| 1.0014844171615094| TRUE        |
| 20       | 2.4907093048095703  | 1.612083911895752| 1.545024602274324| TRUE        |
| 1        | 0.12787652015686035 | 0.12776947021484375| 1.0008378366274557| FALSE       |
| 10       | 0.40375709533691406 | 0.4228525161743164  | 0.9548414160800914  | FALSE       |
| 20       | 2.0639474391937256 | 0.9705793857574463  | 2.129852698601417  | FALSE       |

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

zhuhaozhe committed:
Update on "enable bf16 embeddingbag with fbgemm fast path in aten"


Summary:

This PR adds BF16 support For embedding_bag(CPU) op with FBGEMM kernel
(depends on https://github.com/pytorch/FBGEMM/pull/1540)
 
Benchmark result on CPX 28 core/socket BS = 28 * 128:
```python
import torch

a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)

def cache_flush():
    # We assume the cache size is <= 512MB here.
    # a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # a, b are initialized out of this function to avoid allocate memory every time
    global a, b
    a += b

BATCHSIZE = 128 * 28

E = torch.nn.EmbeddingBag(100000, 512)

import time

for BAGSIZE in [20, 10, 1]:
    input = torch.randint(0, 100000, (BATCHSIZE * BAGSIZE, ))
    offset = torch.arange(0, BATCHSIZE * BAGSIZE, BAGSIZE)
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = E(input, offset)
        t +=time.time() - start
    print("fp32:", t)

    bf16 = E.bfloat16()
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = bf16(input, offset)
        t +=time.time() - start
    print("bf16:",t)
```


| BAG SIZE | FP32               | BF16                | ratio | flush cache |
|----------|--------------------|---------------------|-------|-------------|
| 1        | 0.42674922943115234| 0.4264225959777832| 1.0007659853311013| TRUE        |
| 10       | 1.0786781311035156| 1.0770792961120605| 1.0014844171615094| TRUE        |
| 20       | 2.4907093048095703  | 1.612083911895752| 1.545024602274324| TRUE        |
| 1        | 0.12787652015686035 | 0.12776947021484375| 1.0008378366274557| FALSE       |
| 10       | 0.40375709533691406 | 0.4228525161743164  | 0.9548414160800914  | FALSE       |
| 20       | 2.0639474391937256 | 0.9705793857574463  | 2.129852698601417  | FALSE       |

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

zhuhaozhe commented:
Hi, @jianyuh. May you help to review this PR?

jianyuh commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -305,10 +287,10 @@ index_select_add(
               /*out=*/output_data_fp32 + start_idx * ddim);
           for (int64_t i = start_idx; i < end_idx; i++) {
-            // Convert FP32 intermediate buffer result back to FP16/BF16 for
+            // Convert FP32 intermediate buffer result back to 16bit/BF16 for
```
"16bit/BF16" -> 16bit

jianyuh commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -625,48 +606,51 @@ index_select_scale_add(
+    } else {
       fbgemm::Float16ToFloat_simd(
-          reinterpret_cast<const float16*>(scale_data),
+          reinterpret_cast<const uint16_t*>(scale_data),
```
Use fbgemm::float16?

jianyuh commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -625,48 +606,51 @@ index_select_scale_add(
+    bool isbf16 = std::is_same<data_t, at::Half>::value ? false : true;
+    if (isbf16) {
+      fbgemm::Bfloat16ToFloat_simd(
+          reinterpret_cast<const uint16_t*>(scale_data),
```
uint16_t -> fbgemm::bfloat16 ?

jianyuh commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -716,62 +700,61 @@ index_select_scale_add(
-      (output_data + output_stride0 * i)[d * output_stride1] =
-          static_cast<data_t>((output_data_fp32 + ddim * i)[d]);
+    for (const auto i : c10::irange(output.size(0))) {
+      // Convert FP32 intermediate buffer result back to 16bit/BF16 for output
```
"16bit/BF16" -> "FP16/BF16" ?

zhuhaozhe committed:
Update on "enable bf16 embeddingbag with fbgemm fast path in aten"


Summary:

This PR adds BF16 support For embedding_bag(CPU) op with FBGEMM kernel
(depends on https://github.com/pytorch/FBGEMM/pull/1540)
 
Benchmark result on CPX 28 core/socket BS = 28 * 128:
```python
import torch

a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)

def cache_flush():
    # We assume the cache size is <= 512MB here.
    # a = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # b = torch.ones(256 * 1024 * 1024 // 4, dtype=torch.float)
    # a, b are initialized out of this function to avoid allocate memory every time
    global a, b
    a += b

BATCHSIZE = 128 * 28

E = torch.nn.EmbeddingBag(100000, 512)

import time

for BAGSIZE in [20, 10, 1]:
    input = torch.randint(0, 100000, (BATCHSIZE * BAGSIZE, ))
    offset = torch.arange(0, BATCHSIZE * BAGSIZE, BAGSIZE)
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = E(input, offset)
        t +=time.time() - start
    print("fp32:", t)

    bf16 = E.bfloat16()
    t = 0
    for i in range(1000):
        cache_flush()
        start = time.time()
        y = bf16(input, offset)
        t +=time.time() - start
    print("bf16:",t)
```


| BAG SIZE | FP32               | BF16                | ratio | flush cache |
|----------|--------------------|---------------------|-------|-------------|
| 1        | 0.42674922943115234| 0.4264225959777832| 1.0007659853311013| TRUE        |
| 10       | 1.0786781311035156| 1.0770792961120605| 1.0014844171615094| TRUE        |
| 20       | 2.4907093048095703  | 1.612083911895752| 1.545024602274324| TRUE        |
| 1        | 0.12787652015686035 | 0.12776947021484375| 1.0008378366274557| FALSE       |
| 10       | 0.40375709533691406 | 0.4228525161743164  | 0.9548414160800914  | FALSE       |
| 20       | 2.0639474391937256 | 0.9705793857574463  | 2.129852698601417  | FALSE       |

cc jgong5 mingfeima XiaobingSuper sanchitintel ashokei jingxu10

[ghstack-poisoned]

zhuhaozhe commented:
Hi, @jianyuh. Thanks for your review, I have resolved your comment. May you also help to review https://github.com/pytorch/pytorch/pull/89199?

jianyuh commented on aten/src/ATen/native/EmbeddingBag.cpp:
```diff
@@ -247,44 +233,40 @@ index_select_add(
+              /* prefetch */ 16,
+              /* is_weight_positional */ false,
+              /* use_offsets */ true,
+              /* isbf16*/ isbf16);
```
"/* isbf16 */"

jianyuh commented on CMakeLists.txt:
```diff
@@ -794,10 +794,6 @@ if(USE_COREML_DELEGATE)
   string(APPEND CMAKE_CXX_FLAGS " -DUSE_COREML_DELEGATE")
 endif()
 
-if(BUILD_CAFFE2)
```
Hmm, this is reverting the changes in https://github.com/pytorch/pytorch/pull/89199 . Maybe change the stacked PR to avoid such changes?

zhuhaozhe commented:
Close with https://github.com/pytorch/pytorch/pull/94163.
